### PR TITLE
Implemented Table, Column and Cell

### PR DESCRIPTION
**Pull request by henriette-einstein**
Added 

* Table
* Column
* Cell

and some test cases

### CODE REVIEW

File: packages/core/spec/share/asciidoctor-spec.js [hunk 1]
```diff
@@ -1665,6 +1665,117 @@ America/New_York
         expect(html).to.include('<style>pre.prism{background-color: lightgrey}</style>')
       })
     })
+    describe('Table stuff', function () {
+      it('should return true for hasRows()', function () {
+        const options = {}
+        const doc = asciidoctor.load('= Document Title\n\n|===\n\n|This is a title\n|===', options)
```
Comment (Mogztter) [inline packages/core/spec/share/asciidoctor-spec.js]: Could you please use template literals? I find it easier to read.

File: packages/core/src/asciidoctor-core-api.js [hunk 2]
```diff
@@ -3047,3 +3047,402 @@ Opal.Asciidoctor.SyntaxHighlighterBase = SyntaxHighlighterBase
...
+Table.create = function (parent, attributes) {
+  return this.$new(parent, attributes)
+}
+
+/**
+ * Check, if a given option is set
+ * @param {string} opt the option to check
+ * @returns {boolean}
+ */
+Table.prototype.hasOption = function (opt) {
```
Comment (Mogztter) [inline packages/core/src/asciidoctor-core-api.js]: As a first step I would like to map only existing methods from Asciidoctor core and then we can add utility functions like this one.
Could you please move utility functions in another pull request?

File: packages/core/src/asciidoctor-core-api.js [hunk 3]
```diff
@@ -3047,3 +3047,402 @@ Opal.Asciidoctor.SyntaxHighlighterBase = SyntaxHighlighterBase
...
+/**
+ * Create a new Column element
+ * @param {Table} table
+ * @param {number} index
+ * @param {Object|undefined} attributes
+ * @returns Column
+ */
+Column.create = function (table, index, attributes) {
+  let ret = this.$new(table, index, attributes)
+  return ret
```
Comment (Mogztter) [inline packages/core/src/asciidoctor-core-api.js]: It could be inlined right?

```suggestion
  return this.$new(table, index, attributes)
```

File: packages/core/src/asciidoctor-core-api.js [hunk 4]
```diff
@@ -3047,3 +3047,402 @@ Opal.Asciidoctor.SyntaxHighlighterBase = SyntaxHighlighterBase
...
+
+/**
+ * Create a new Column element
+ * @param {Table} table
+ * @param {number} index
+ * @param {Object|undefined} attributes
+ * @returns Column
+ */
+Column.create = function (table, index, attributes) {
+  let ret = this.$new(table, index, attributes)
```
Comment (Mogztter) [inline packages/core/src/asciidoctor-core-api.js]: I think we should call `toHash` on attributes.

File: packages/core/src/asciidoctor-core-api.js [hunk 5]
```diff
@@ -3047,3 +3047,402 @@ Opal.Asciidoctor.SyntaxHighlighterBase = SyntaxHighlighterBase
...
+  return this.getAttribute('valign')
+}
+
+/**
+ * Get the style of this cell
+ * @returns {string}
+ */
+Column.prototype.getStyle = function () {
+  let ret = this.style
+  return (ret !== Opal.nil) ? ret : undefined
```
Comment (Mogztter) [inline packages/core/src/asciidoctor-core-api.js]: Could you please use the following format (for consistency):

```suggestion
  var style = this.style
  return style === Opal.nil ? undefined : style
```

So:
- `var` instead of `let` (we might use `let` and `const` soon but for now we stick with `var` mostly to maintain compatibility)
- use the name of the method as the name of the variable (`this.style` -> `var style`)
- compare the variable to `Opal.nil` and return undefined or the value (`style === Opal.nil ? undefined : style`) without parenthesis.

File: settings.json [hunk 6]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (Mogztter) [inline settings.json]: I don't mind adding this file. Is that a Visual Studio Code settings? Is it safe? (ie. I don't want to override user-specific settings or force a specific settings)

**Comment by Mogztter** [comment]
Overall that's look very good :+1: 
I left a few comments.

We also maintain an handcrafted TypeScript Definition File: https://github.com/asciidoctor/asciidoctor.js/blob/master/packages/core/types/index.d.ts

So we will need to add the new functions in here too. Let me know if you are not familiar with TypeScript or if you need help to get started. I can also follow-up on your work and add the missing definition in the `index.d.ts` file.

File: settings.json (hunk 6 above)
Comment (henriette-einstein) [reply settings.json]: The settings are defaults, if you use the javascript standard lint rules. However, I would not merge this, since other users may have different settings

File: packages/core/src/asciidoctor-core-api.js (hunk 3 above)
Comment (henriette-einstein) [reply packages/core/src/asciidoctor-core-api.js]: Done. Will be included in the next commit

File: packages/core/src/asciidoctor-core-api.js (hunk 4 above)
Comment (henriette-einstein) [reply packages/core/src/asciidoctor-core-api.js]: Done. Will be included in the next commit

File: packages/core/src/asciidoctor-core-api.js (hunk 5 above)
Comment (henriette-einstein) [reply packages/core/src/asciidoctor-core-api.js]: That may be in more than one place. I'll check it before committing

File: packages/core/spec/share/asciidoctor-spec.js (hunk 1 above)
Comment (henriette-einstein) [reply packages/core/spec/share/asciidoctor-spec.js]: Done. Will be included in the next commit

File: packages/core/src/asciidoctor-core-api.js (hunk 2 above)
Comment (henriette-einstein) [reply packages/core/src/asciidoctor-core-api.js]: Done. Will be included in the next commit

**Comment by henriette-einstein** [comment]
Sorry, I am not familiar with the TypeScript definition files

**Comment by henriette-einstein** [comment]
@Mogztter I have no idea, why the build

Build / build (ubuntu-latest, 13.x, master) (pull_request)

failed. What can I do?

**Comment by Mogztter** [comment]
> Sorry, I am not familiar with the TypeScript definition files

No worries, I will take care of it 👍

> I have no idea, why the build
Build / build (ubuntu-latest, 13.x, master) (pull_request)
failed. What can I do?

I think it's an intermittent failure, I might need to slightly increase the timeout. Nothing to worry about but now GitHub actions does not find the commit reference 🤔 
Can you please force push to trigger a new build?

File: settings.json (hunk 6 above)
Comment (Mogztter) [reply settings.json]: Ok, in this case can you please remove it?

**Comment by Mogztter** [comment]
Build is green, nice 🎉 
If that's OK with you, I will add commits on top of your work to adjust a few things and add the type definition.

**Comment by henriette-einstein** [comment]
That is fine with me.

**Comment by danyill** [comment]
Thanks @henriette-einstein, thanks @Mogztter much appreciated.

File: packages/core/src/asciidoctor-core-api.js [hunk 7]
```diff
@@ -3047,3 +3047,381 @@ Opal.Asciidoctor.SyntaxHighlighterBase = SyntaxHighlighterBase
...
+ */
+Cell.prototype.getWidth = function () {
+  return this.getAttribute('width')
+}
+
+/**
+ * Get the colpcwidth of this cell
+ * @returns {string|undefined}
+ */
+Cell.prototype.getColpcwidth = function () {
```
Comment (Mogztter) [inline packages/core/src/asciidoctor-core-api.js]: I think we should use long form (ie. `getColumnPercentageWidth`) because the rest of the API is using long form.
At some point, we might introduce alias:

- `getColPercentWidth`
- `getColPcWidth`

But we need to be careful because users will expect `Col` to be equivalent to `Col` in other functions as well.

File: packages/core/src/asciidoctor-core-api.js [hunk 8]
```diff
@@ -3047,3 +3047,381 @@ Opal.Asciidoctor.SyntaxHighlighterBase = SyntaxHighlighterBase
...
+Cell.prototype.getLines = function () {
+  var lines = this['$lines']()
+  return lines === Opal.nil ? undefined : lines
+}
+
+/**
+ * Get the lineno of the cell
+ * @returns {string}
+ */
+Cell.prototype.getLineno = function () {
```
Comment (Mogztter) [inline packages/core/src/asciidoctor-core-api.js]: It should be `getLineNumber`

File: packages/core/src/asciidoctor-core-api.js [hunk 9]
```diff
@@ -3047,3 +3047,381 @@ Opal.Asciidoctor.SyntaxHighlighterBase = SyntaxHighlighterBase
...
+  var value = this.getAttributes()[name]
+  return value === Opal.nil ? undefined : value
+}
+
+/**
+ * Get the colspan of this {@link Cell} node
+ * @memberof Cell
+ * @returns {number}  An Integer of the number of columns this cell will span (default undefines)
+ */
+Cell.prototype.getColspan = function () {
```
Comment (Mogztter) [inline packages/core/src/asciidoctor-core-api.js]: If we apply the same rules, then `getColumnSpan`

File: packages/core/src/asciidoctor-core-api.js (hunk 7 above)
Comment (henriette-einstein) [reply packages/core/src/asciidoctor-core-api.js]: Changed name to getColumnPercentageWidth. No alias defined by now

File: packages/core/src/asciidoctor-core-api.js (hunk 9 above)
Comment (henriette-einstein) [reply packages/core/src/asciidoctor-core-api.js]: Changed to getColumnSpan

File: packages/core/src/asciidoctor-core-api.js (hunk 8 above)
Comment (henriette-einstein) [reply packages/core/src/asciidoctor-core-api.js]: Changed to getLineNumber

File: packages/core/src/asciidoctor-core-api.js [hunk 10]
```diff
@@ -3047,3 +3047,387 @@ Opal.Asciidoctor.SyntaxHighlighterBase = SyntaxHighlighterBase
...
+  return this.getAttribute('width')
+}
+
+/**
+ * Get the column width in percentage of this cell.
+ * @returns {string|undefined}
+ * @memberof Cell
+ */
+Cell.prototype.getColumnPercentageWidth = function () {
+  return this.getAttribute('colpcwidth')
```
Comment (Mogztter) [inline packages/core/src/asciidoctor-core-api.js]: @mojavelinux How do you feel about these utilities functions to "quickly" access the underlying attribute?

File: packages/core/src/asciidoctor-core-api.js [hunk 11]
```diff
@@ -3047,3 +3047,387 @@ Opal.Asciidoctor.SyntaxHighlighterBase = SyntaxHighlighterBase
...
+}
+
+/**
+ * Check if the table has the autowidth option set.
+ * @returns {boolean}
+ * @memberof Table
+ */
+Table.prototype.hasAutowidthOption = function () {
+  var autowidthOption = this.getAttributes()['autowidth-option']
+  return autowidthOption === Opal.nil ? false : autowidthOption === ''
```
Comment (Mogztter) [inline packages/core/src/asciidoctor-core-api.js]: @mojavelinux I'm always a bit concerned that someone could use `setAttribute` with a boolean value:

```js
table.setAttribute('autowidth-option', true)
table.hasAutowidthOption() // false?!
```

Do we need to check if the value is truthy instead? 

```suggestion
  return autowidthOption === Opal.nil ? false : autowidthOption
```

File: packages/core/src/asciidoctor-core-api.js [hunk 12]
```diff
@@ -3047,3 +3047,387 @@ Opal.Asciidoctor.SyntaxHighlighterBase = SyntaxHighlighterBase
...
+Table.prototype.getFoot = function () {
+  return this.rows.foot
+}
+
+/**
+ * Check if the table has rows.
+ * @returns {boolean}
+ * @memberof Table
+ */
+Table.prototype.hasRows = function () {
```
Comment (Mogztter) [inline packages/core/src/asciidoctor-core-api.js]: I think it should be `hasBodyRows`. What do you think?

File: packages/core/src/asciidoctor-core-api.js [hunk 13]
```diff
@@ -3047,3 +3047,387 @@ Opal.Asciidoctor.SyntaxHighlighterBase = SyntaxHighlighterBase
...
+Table.prototype.getBody = function () {
+  return this.rows.body
+}
+
+/**
+ * Get the foot rows of this table.
+ * @returns {Array<Array<Cell>>} - an Array of Array of Cell
+ * @memberof Table
+ */
+Table.prototype.getFoot = function () {
```
Comment (Mogztter) [inline packages/core/src/asciidoctor-core-api.js]: `getFootRows`

File: packages/core/src/asciidoctor-core-api.js [hunk 14]
```diff
@@ -3047,3 +3047,387 @@ Opal.Asciidoctor.SyntaxHighlighterBase = SyntaxHighlighterBase
...
+Table.prototype.hasRows = function () {
+  return this.rows !== Opal.nil && this.rows.body.length > 0
+}
+
+/**
+ * Check if the table has a header.
+ * @returns {boolean}
+ * @memberof Table
+ */
+Table.prototype.hasHeader = function () {
```
Comment (Mogztter) [inline packages/core/src/asciidoctor-core-api.js]: `getHeadRows` ?

File: packages/core/src/asciidoctor-core-api.js (hunk 12 above)
Comment (henriette-einstein) [reply packages/core/src/asciidoctor-core-api.js]: You are right. I'll change that

File: packages/core/src/asciidoctor-core-api.js (hunk 13 above)
Comment (henriette-einstein) [reply packages/core/src/asciidoctor-core-api.js]: getFootRows is definitely better. I'll change it

File: packages/core/src/asciidoctor-core-api.js [hunk 15]
```diff
@@ -3047,3 +3047,387 @@ Opal.Asciidoctor.SyntaxHighlighterBase = SyntaxHighlighterBase
...
+ */
+Table.prototype.hasHeaderOption = function () {
+  return this.has_header_option
+}
+
+/**
+ * Check if the table has a footer.
+ * @returns {boolean}
+ */
+Table.prototype.hasFooter = function () {
```
Comment (Mogztter) [inline packages/core/src/asciidoctor-core-api.js]: I think we should be consistent with the naming.
Here the underlying property is `foot` so `hasFoot`?

@mojavelinux What is the reasoning behind `foot` and `head` (instead of footer` and `header`)?

File: packages/core/src/asciidoctor-core-api.js (hunk 7 above)
Comment (Mogztter) [reply packages/core/src/asciidoctor-core-api.js]: Thanks!

File: packages/core/src/asciidoctor-core-api.js (hunk 15 above)
Comment (henriette-einstein) [reply packages/core/src/asciidoctor-core-api.js]: It is probably due to the fact, that in HTML the corresponding tags are 'head', 'body' and 'foot'

File: packages/core/src/asciidoctor-core-api.js (hunk 14 above)
Comment (henriette-einstein) [reply packages/core/src/asciidoctor-core-api.js]: I'll change that and also hasFooter to hasFootRows

**Comment by Mogztter** [comment]
I did a little cleanup, it's looking very good.
I want to hear Dan's opinion on a few specific cases and then I think we will be ready to move forward!

Thanks again for the heavy lifting :muscle:

**Comment by henriette-einstein** [comment]
Thanks for the great project. You did the heavy work! I'll add some more tascaseos and will commit in a few minutes

File: packages/core/src/asciidoctor-core-api.js (hunk 15 above)
Comment (Mogztter) [reply packages/core/src/asciidoctor-core-api.js]: That's a good point but I wonder if it would do more harm than good in the API (ie. it can be confusing to use `hasHeader` but `getHead`).

File: packages/core/src/asciidoctor-core-api.js (hunk 14 above)
Comment (henriette-einstein) [reply packages/core/src/asciidoctor-core-api.js]: I'll also add a method hasBodyRows

**Comment by Mogztter** [comment]
> Thanks for the great project. You did the heavy work!

:blush: 
Then a team effort!

> I'll add some more test cases and will commit in a few minutes

Perfect :ok_hand:

File: packages/core/src/asciidoctor-core-api.js (hunk 15 above)
Comment (henriette-einstein) [reply packages/core/src/asciidoctor-core-api.js]: The methods are now renamed to hasHeadRows, hasBodyRows and hasFootRows

**Comment by Mogztter** [comment]
I've rebased on master (force-push), added the TypeScript definitions and a changelog.
I've also fix a few inconsistencies/differences.
It should be ready to go!
